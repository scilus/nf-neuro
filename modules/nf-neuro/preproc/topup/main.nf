process PREPROC_TOPUP {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        "https://scil.usherbrooke.ca/containers/scilus_2.0.2.sif":
        "scilus/scilus:2.0.2"}"

    input:
        tuple val(meta), path(dwi), path(bval), path(bvec), path(b0), path(rev_dwi), path(rev_bval), path(rev_bvec), path(rev_b0)
        val(config_topup)

    output:
        tuple val(meta), path("*__corrected_b0s.nii.gz"), emit: topup_corrected_b0s
        tuple val(meta), path("*_fieldcoef.nii.gz")     , emit: topup_fieldcoef
        tuple val(meta), path("*_movpar.txt")           , emit: topup_movpart
        tuple val(meta), path("*__rev_b0_warped.nii.gz"), emit: rev_b0_warped
        tuple val(meta), path("*__rev_b0_mean.nii.gz")  , emit: rev_b0_mean
        tuple val(meta), path("*__b0_mean.nii.gz")      , emit: b0_mean
        path "versions.yml"                             , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"

    def prefix_topup = task.ext.prefix_topup ? task.ext.prefix_topup : ""
    def config_topup = config_topup ?: task.ext.default_config_topup
    def encoding = task.ext.encoding ? task.ext.encoding : ""
    def readout = task.ext.readout ? task.ext.readout : ""
    def b0_thr_extract_b0 = task.ext.b0_thr_extract_b0 ? task.ext.b0_thr_extract_b0 : ""

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1
    export ANTS_RANDOM_SEED=7468

    if [[ -f "$b0" ]];
    then
        scil_volume_math.py concatenate $b0 $b0 ${prefix}__concatenated_b0.nii.gz --data_type float32
        scil_volume_math.py mean ${prefix}__concatenated_b0.nii.gz ${prefix}__b0_mean.nii.gz
    else
        scil_dwi_extract_b0.py $dwi $bval $bvec ${prefix}__b0_mean.nii.gz --mean --b0_threshold $b0_thr_extract_b0 --skip_b0_check
    fi

    if [[ -f "$rev_b0" ]];
    then
        scil_volume_math.py concatenate $rev_b0 $rev_b0 ${prefix}__concatenated_rev_b0.nii.gz --data_type float32
        scil_volume_math.py mean ${prefix}__concatenated_rev_b0.nii.gz ${prefix}__rev_b0_mean.nii.gz
    else
        scil_dwi_extract_b0.py $rev_dwi $rev_bval $rev_bvec ${prefix}__rev_b0_mean.nii.gz --mean --b0_threshold $b0_thr_extract_b0 --skip_b0_check
    fi

    antsRegistrationSyNQuick.sh -d 3 -f ${prefix}__b0_mean.nii.gz -m ${prefix}__rev_b0_mean.nii.gz -o output -t r -e 1
    mv outputWarped.nii.gz ${prefix}__rev_b0_warped.nii.gz
    scil_dwi_prepare_topup_command.py ${prefix}__b0_mean.nii.gz ${prefix}__rev_b0_warped.nii.gz\
        --config $config_topup\
        --encoding_direction $encoding\
        --readout $readout --out_prefix $prefix_topup\
        --out_script
    sh topup.sh
    cp corrected_b0s.nii.gz ${prefix}__corrected_b0s.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
        antsRegistration: \$(antsRegistration --version | grep "Version" | sed -E 's/.*v([0-9]+\\+\\).*/\\1/')
        fsl: \$(flirt -version 2>&1 | sed -n 's/FLIRT version \\([0-9.]\\+\\)/\\1/p')

    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def prefix_topup = task.ext.prefix_topup ? task.ext.prefix_topup : ""

    """
    touch ${prefix}__corrected_b0s.nii.gz
    touch ${prefix}__rev_b0_warped.nii.gz
    touch ${prefix}__rev_b0_mean.nii.gz
    touch ${prefix}__b0_mean.nii.gz
    touch ${prefix_topup}_fieldcoef.nii.gz
    touch ${prefix_topup}_movpar.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(pip list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
        antsRegistration: \$(antsRegistration --version | grep "Version" | sed -E 's/.*v([0-9]+\\+\\).*/\\1/')
        fsl: \$(flirt -version 2>&1 | sed -n 's/FLIRT version \\([0-9.]\\+\\)/\\1/p')
    END_VERSIONS

    function handle_code () {
    local code=\$?
    ignore=( 1 )
    exit \$([[ " \${ignore[@]} " =~ " \$code " ]] && echo 0 || echo \$code)
    }
    trap 'handle_code' ERR

    scil_volume_math.py -h
    scil_dwi_extract_b0.py -h
    antsRegistrationSyNQuick.sh
    scil_dwi_prepare_topup_command.py -h
    """
}
