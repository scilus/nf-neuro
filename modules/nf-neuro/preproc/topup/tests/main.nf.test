nextflow_process {

    name "Test Process PREPROC_TOPUP"
    script "../main.nf"
    process "PREPROC_TOPUP"

    tag "modules"
    tag "modules_nfcore"
    tag "preproc"
    tag "preproc/topup"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    config "./nextflow.config"

    setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "topup_eddy_light.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("topup_full") {
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory
                    .map{ test_data_directory -> [
                    [ id:'test', single_end:false ], // meta map
                    file("\${test_data_directory}/sub-01_dir-AP_dwi.nii.gz", checkIfExists: true),
                    file("\${test_data_directory}/sub-01_dir-AP_dwi.bval", checkIfExists: true),
                    file("\${test_data_directory}/sub-01_dir-AP_dwi.bvec", checkIfExists: true),
                    file("\${test_data_directory}/sub-01_dir-AP_sbref.nii.gz", checkIfExists: true),
                    file("\${test_data_directory}/sub-01_dir-PA_dwi.nii.gz", checkIfExists: true),
                    file("\${test_data_directory}/sub-01_dir-PA_dwi.bval", checkIfExists: true),
                    file("\${test_data_directory}/sub-01_dir-PA_dwi.bvec", checkIfExists: true),
                    file("\${test_data_directory}/sub-01_dir-PA_sbref.nii.gz", checkIfExists: true)]}
                input[1] = []
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    file(process.out.topup_corrected_b0s.get(0).get(1)).name,
                    file(process.out.topup_fieldcoef.get(0).get(1)).name,
                    file(process.out.topup_movpart.get(0).get(1)).name,
                    file(process.out.rev_b0_warped.get(0).get(1)).name,
                    process.out.rev_b0_mean,
                    process.out.b0_mean,
                    process.out.versions
                ).match() }
            )
        }
    }

    test("topup_light") {
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory
                    .map{ test_data_directory -> [
                    [ id:'test', single_end:false ], // meta map
                    file("\${test_data_directory}/sub-01_dir-AP_dwi.nii.gz", checkIfExists: true),
                    file("\${test_data_directory}/sub-01_dir-AP_dwi.bval", checkIfExists: true),
                    file("\${test_data_directory}/sub-01_dir-AP_dwi.bvec", checkIfExists: true),
                    [],
                    [],
                    [],
                    [],
                    file("\${test_data_directory}/sub-01_dir-PA_sbref.nii.gz", checkIfExists: true)]}
                input[1] = []
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    file(process.out.topup_corrected_b0s.get(0).get(1)).name,
                    file(process.out.topup_fieldcoef.get(0).get(1)).name,
                    file(process.out.topup_movpart.get(0).get(1)).name,
                    file(process.out.rev_b0_warped.get(0).get(1)).name,
                    process.out.rev_b0_mean,
                    process.out.b0_mean,
                    process.out.versions
                ).match() }
            )
        }
    }
}
