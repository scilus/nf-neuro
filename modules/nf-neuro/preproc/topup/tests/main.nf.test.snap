{
    "topup_full": {
        "content": [
            "test__corrected_b0s.nii.gz",
            "topup_results_fieldcoef.nii.gz",
            "topup_results_movpar.txt",
            "test__rev_b0_warped.nii.gz",
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__rev_b0_mean.nii.gz:md5,a0c63d7a8a234d85eeb63045ae2a8f60"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__b0_mean.nii.gz:md5,f50e652bcdd1d45f7a4c592df7eb851a"
                ]
            ],
            [
                "versions.yml:md5,87d6001e63fd82f49d6dd80ed032d4c0"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-11-01T16:21:18.752887499"
    },
    "topup_light": {
        "content": [
            "test__corrected_b0s.nii.gz",
            "topup_results_fieldcoef.nii.gz",
            "topup_results_movpar.txt",
            "test__rev_b0_warped.nii.gz",
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__rev_b0_mean.nii.gz:md5,a0c63d7a8a234d85eeb63045ae2a8f60"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__b0_mean.nii.gz:md5,91288914a80e7866ef8b08a363d617f9"
                ]
            ],
            [
                "versions.yml:md5,87d6001e63fd82f49d6dd80ed032d4c0"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-11-01T16:21:36.189599346"
    }
}